l5kit/l5kit/evaluation/csv_utils.py

```diff
@@ -0,0 +1,246 @@
+        track_ids (np.ndarray): (num_example,) agent ids
+        coords (np.ndarray): (num_example x future_len x num_coords) meters displacements
+        avails (np.ndarray): (num_example x future_len) array with value 0 (discard in eval) or 1 (keep in eval)
+
+    Returns:
+
+    """
+    assert len(coords.shape) == 3
+    num_example, future_len, num_coords = coords.shape
+    assert num_coords == 2
+    assert timestamps.shape == track_ids.shape == (num_example,)
+    assert avails.shape == (num_example, future_len)
+
+    coords_keys = _generate_coords_keys(future_len)
+    avails_keys = _generate_avails_keys(future_len)
+
+    # create and write HEADER
+    # order is (timestamp,track_id,availabilities,coords)
+    fieldnames = ["timestamp", "track_id"]
+    fieldnames.extend(avails_keys)
+    fieldnames.extend(coords_keys)
+    writer = csv.DictWriter(open(csv_path, "w"), fieldnames)
+    writer.writeheader()
+
+    for timestamp, track_id, coord, avail in zip(timestamps, track_ids, coords, avails):
+        # writing using dict relieves us from respecting the order
+        line = {"timestamp": timestamp, "track_id": track_id}
+        line.update({key: ava for key, ava in zip(avails_keys, avail)})
+        # TODO limit digits in coords?
+        line.update({key: cor for key, cor in zip(coords_keys, coord.reshape(-1))})
+
+        writer.writerow(line)
+
+
+def read_gt_csv(csv_path: str) -> Iterator[dict]:
+    """
+    Generator function that returns a line at the time from the csv file as a dict
+
+    Args:
+        csv_path (str): path of the csv to read
+
+    Returns:
+        Iterator[dict]: dict keys are the csv header fieldnames
+    """
+    reader = csv.DictReader(open(csv_path, "r"))
+    fieldnames = reader.fieldnames
+    assert fieldnames is not None, "error reading fieldnames"
+
+    future_len = (len(fieldnames) - 2) / 3  # excluded first two fields, the rest must be (x, y, av) * len = 3*len
+    assert future_len == int(future_len), "error estimating len"
+    future_len = int(future_len)
+
+    coords_keys = _generate_coords_keys(future_len)
+    avails_keys = _generate_avails_keys(future_len)
+
+    for row in reader:
+        timestamp = row["timestamp"]
+        track_id = row["track_id"]
+
+        avail = np.asarray([np.float64(row[key]) for key in avails_keys])
+
+        coord = np.asarray([np.float64(row[key]) for key in coords_keys])
+        coord = coord.reshape((future_len, 2))
+
+        yield {"track_id": track_id, "timestamp": timestamp, "coord": coord, "avail": avail}
+
+
+def write_pred_csv(
+    csv_path: str,
+    timestamps: np.ndarray,
+    track_ids: np.ndarray,
+    coords: np.ndarray,
+    confs: Optional[np.ndarray] = None,
+) -> None:
+    """
+    Encode the ground truth into a csv file. Coords can have an additional axis for multi-mode.
+    We handle up to MAX_MODES modes.
+    For the uni-modal case, coords should not have the additional axis and confs should be set to None.
```

Comment:
Minor: "For the uni-modal case' sounds a bit like there can be multi-modal models which only output a single prediction for certain umabigious cases, and in this case should fall back to outputting the uni-modal format (whereas, we just distinguish between models which always output uni-modal predicitions and models which always output multi-modal predictions). With a good explanation of this e.g. in the competition this should be ok though.

Reply:
I've added a comment to ensure it's clear that every prediction must be uni-modal

Reply:
TBH you could have a mixed prediction set where some predictions have just a single mode with confidence set to 1. In the end, uni-modal is just a special case of multi :). The unimodal option is to avoid to specify confidences at all really.

You're right, we should clarify it better in the competiion